lib/spack/spack/spec.py

```diff
@@ -4995,53 +4995,74 @@ def __missing__(self, key):
 
 #: These are possible token types in the spec grammar.
 HASH, DEP, VER, COLON, COMMA, ON, OFF, PCT, EQ, ID, VAL, FILE = range(12)
+#  0,   1,   2,     3,     4,  5,   6,   7,  8,  9,  10,   11
 
 #: Regex for fully qualified spec names. (e.g., builtin.hdf5)
 spec_id_re = r"\w[\w.-]*"
 
 
+_filename_re_base = r"[/\w.-]*/[/\w/-]+\.(yaml|json)[^\b]*"
+_windows_filename_prefix = r"([A-Za-z]:)*?"
+# Spec strings require posix-style paths on Windows
+# because the result is later passed to shlex.
+_windows_filename_re = _windows_filename_prefix + _filename_re_base
+
+#: Regex for literal filenames.
+filename_reg = _windows_filename_re if is_windows else _filename_re_base
+
+
 class SpecLexer(spack.parse.Lexer):
 
     """Parses tokens that make up spack specs."""
 
+    # TODO: Determine whether using >2 lexer modes would allow us to create
+    # a separate entry (likely defined the same as `spec_id_re` at first) to
+    # specifically match version strings, variant names, compiler dependency
+    # names (with '%').
     def __init__(self):
-        # Spec strings require posix-style paths on Windows
-        # because the result is later passed to shlex
-        filename_reg = (
-            r"[/\w.-]*/[/\w/-]+\.(yaml|json)[^\b]*"
-            if not is_windows
-            else r"([A-Za-z]:)*?[/\w.-]*/[/\w/-]+\.(yaml|json)[^\b]*"
-        )
         super(SpecLexer, self).__init__(
             [
                 (
-                    r"\@([\w.\-]*\s*)*(\s*\=\s*\w[\w.\-]*)?",
-                    lambda scanner, val: self.token(VER, val),
+                    "BEGIN_PHASE",
+                    [
+                        # '@': begin a Version, VersionRange, or VersionList:
+                        (r"@([\w.\-]*\s*)*(\s*\=\s*\w[\w.\-]*)?", VER),
```

Comment:
The only actual change I made to any of these strings is to remove the backslash from `r"\@"`  here since the backslash is simply ignored for any non-metacharacters.